File: python/2017/day1.py


def solve_captcha(line):
    """
    >>> solve_captcha('1122')
    3
    >>> solve_captcha('1111')
    4
    >>> solve_captcha('1234')
    0
    >>> solve_captcha('91212129')
    9
    """
    return sum(int(c) for i, c in enumerate(line) if line[i-1] == c)


def part1(lines):
    solve_captcha(lines[0].strip())


def solve_new_captcha(line):
    """
    >>> solve_new_captcha('1212')
    6
    >>> solve_new_captcha('1221')
    0
    >>> solve_new_captcha('123425')
    4
    >>> solve_new_captcha('123123')
    12
    >>> solve_new_captcha('12131415')
    4
    """
    return sum(int(x) for x, y in zip(line, line[len(line)//2:] + line[:len(line)//2]) if x == y)


def part2(lines):
    part2(lines[0].strip())

File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2024/Day1.kt
package io.github.ocirne.aoc.year2024

import io.github.ocirne.aoc.AocChallenge
import kotlin.math.abs

class Day1(val lines: List<String>) : AocChallenge(2024, 1) {

    private val whitespace = Regex("\\s+")

    private fun extractList(index: Int): List<Int> {
        return lines.map { line -> line.split(whitespace)[index].toInt() }.sorted()
    }

    private val leftList = extractList(0)
    private val rightList = extractList(1)

    override fun part1(): Int {
        return leftList.zip(rightList).sumOf { (l, r) -> abs(r - l) }
    }

    override fun part2(): Int {
        val rightCounts = rightList.groupingBy { it }.eachCount()
        return leftList.sumOf { x -> x * rightCounts.getOrDefault(x, 0) }
    }
}
